src/main/java/com/fasterxml/jackson/core/util/ObjectPool.java

```diff
@@ -0,0 +1,330 @@
+            }
+        }
+
+        @Override
+        public void close() throws Exception {
+
+        }
+    }
+
+    class SyncedArrayPool<T> implements ObjectPool<T> {
+        private final Supplier<T> factory;
+
+        private final int size;
+
+        private final Object[] pool;
+
+        private final ReentrantLock lock = new ReentrantLock();
+
+        private volatile int counter = 0;
+
+        public SyncedArrayPool(Supplier<T> factory, int size) {
+            this.factory = factory;
+            this.size = size;
+            this.pool = new Object[size];
+        }
+
+        @Override
+        public T borrow() {
+            lock.lock();
+            try {
+                if (counter == 0) {
+                    return factory.get();
+                }
+                T t = (T) pool[--counter];
+                pool[counter] = null;
+                return t;
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        public void offer(T t) {
+            lock.lock();
+            try {
+                if (counter < size) {
+                    pool[counter++] = t;
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        public void close() throws Exception {
+
+        }
+    }
+
+    class ShardedSyncedArrayPool<T> implements ObjectPool<T> {
+        private final SyncedArrayPool<T>[] pools;
+
+        public ShardedSyncedArrayPool(Supplier<T> factory, int shardsNr, int shardSize) {
+            this.pools = new SyncedArrayPool[shardsNr];
+            for (int i = 0; i < pools.length; i++) {
+                pools[i] = new SyncedArrayPool(factory, shardSize);
+            }
+        }
+
+        @Override
+        public T borrow() {
+            return pools[Thread.currentThread().hashCode() % pools.length].borrow();
+        }
+
+        @Override
+        public void offer(T t) {
+            pools[Thread.currentThread().hashCode() % pools.length].offer(t);
+        }
+
+        @Override
+        public void close() throws Exception {
+
+        }
+    }
+
+    class DummyPool<T> implements ObjectPool<T> {
+        private final Supplier<T> factory;
+
+
+        public DummyPool(Supplier<T> factory) {
+            this.factory = factory;
+        }
+
+        @Override
+        public T borrow() {
+            return factory.get();
+        }
+
+        @Override
+        public void offer(T t) {
+
+        }
+
+        @Override
+        public void close() throws Exception {
+
+        }
+    }
+
+    class JCToolsPool<T> implements ObjectPool<T> {
+        private final MpmcArrayQueue<T> queue;
+
+        private final Supplier<T> factory;
+
+        public JCToolsPool(Supplier<T> factory, int size) {
+            this.factory = factory;
+            this.queue = new MpmcArrayQueue<>(size);
+        }
+
+        @Override
+        public T borrow() {
+            T t = queue.poll();
```

Comment:
you can use https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java#L330 and https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java#L460 to deal with contention in a similar way (and still symmetric) then the lock free pool; if you want to be fair among impls that have this option